csrc/dynamic_transform.cpp

```diff
@@ -270,80 +352,88 @@ void DynamicTransformConcretizer::mutate(TensorView* tv) {
+          tv->fusion(),
+          {tv->getRootDomain().begin(), tv->getRootDomain().end()},
+          {tv->getMaybeRFactorDomain().begin(),
+           tv->getMaybeRFactorDomain().end()});
+      for (auto expr : all_id_exprs) {
+        // Assume outputs of IterDomain exprs are always IterDomains. If
+        // the assumption is invalidated, the logic here would need to
+        // be updated. Assert the assumption to immediately detect such
+        // a case if happened.
+        for (auto out_val : expr->outputs()) {
+          TORCH_INTERNAL_ASSERT(
+              out_val->isA<IterDomain>(),
+              "Unexpected output: ",
+              out_val->toString(),
+              ". IterDomain was expected.");
+        }
+
+        // If any output ID is symbolic, all output IDs should be symbolic
+        if (std::any_of(
+                expr->outputs().begin(),
+                expr->outputs().end(),
+                [](Val* output) {
+                  return output->as<IterDomain>()->getIterType() ==
+                      IterType::Symbolic;
+                })) {
+          TORCH_INTERNAL_ASSERT(std::all_of(
               expr->outputs().begin(), expr->outputs().end(), [](Val* output) {
-                return output->as<IterDomain>()->getIterType() !=
+                return output->as<IterDomain>()->getIterType() ==
                     IterType::Symbolic;
-              })) {
-        continue;
-      }
-      // If any of output IDs is symbolic, all outputs should be symbolic
-      TORCH_INTERNAL_ASSERT(std::all_of(
-          expr->outputs().begin(), expr->outputs().end(), [](Val* output) {
-            return output->as<IterDomain>()->getIterType() ==
-                IterType::Symbolic;
-          }));
-
-      // Determine the output IterType
-      IterType iter_type = IterType::Symbolic;
-      for (auto inp_id : ir_utils::filterByType<IterDomain>(expr->inputs())) {
-        auto updated_id = maybeMutated(inp_id)->as<IterDomain>();
-        iter_type = ops::promoteIterType(iter_type, updated_id->getIterType());
-      }
-      TORCH_INTERNAL_ASSERT(
-          iter_type != IterType::Symbolic,
-          "Failed to concretize an output IterType for expression: ",
-          expr->toString());
-
-      // Update the IterType of each output
-      for (auto out_id : ir_utils::filterByType<IterDomain>(expr->outputs())) {
-        auto concreteized_out_id =
-            IterDomainBuilder(out_id).iter_type(iter_type).build();
-        registerMutation(out_id, concreteized_out_id);
+              }));
+        } else if (std::all_of(
+                       expr->inputs().begin(),
+                       expr->inputs().end(),
+                       [](Val* output) {
+                         return output->as<IterDomain>()->getIterType() !=
+                             IterType::Symbolic;
+                       })) {
+          // If no inputs or outputs are symbolic, nothing to concretize
+          continue;
+        }
```

Comment:
@naoyam this change was necessary to fix failing test `DynamicTransform5_CUDA` which contains a dynamic reshape followed by a static pad (constant pad widths of 1, 1). The pad is inferred to be output itertype `Iteration` at definition, even though its input is symbolic due to the dynamic reshape. That means that concretization needs to handle a symbolic root domain with already-concrete rfactor domain. Previously, skipped mutation in this function if none of the ID expr outputs was symbolic. Now we skip if neither of inputs or outputs is symbolic instead. I preserved the check that if any output is symbolic then they all must be.